File: backend/src/main/java/com/team05/petmeeting/domain/user/service/MailService.java
package com.team05.petmeeting.domain.user.service;

import jakarta.mail.internet.MimeMessage;
import lombok.RequiredArgsConstructor;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.mail.javamail.MimeMessageHelper;
import org.springframework.stereotype.Service;
import org.thymeleaf.TemplateEngine;
import org.thymeleaf.context.Context;

@Service
@RequiredArgsConstructor
public class MailService {

    private final JavaMailSender mailSender;
    private final TemplateEngine templateEngine;

    public void sendMail(String email, String code) {

        MimeMessage message = mailSender.createMimeMessage();

        try {

            MimeMessageHelper helper = new MimeMessageHelper(message, true, "UTF-8");

            helper.setTo(email);
            helper.setSubject("Pet Meeting 이메일 인증 코드");

            Context context = new Context();
            context.setVariable("code", code);

            String html = templateEngine.process("mail/signupotp", context);

            helper.setText(html, true);
            helper.setFrom("no-reply@petmeeting.com", "PetMeeting");

            mailSender.send(message);

        } catch (Exception e) {
            throw new RuntimeException("메일 발송 실패", e);
        }
    }
}

File: backend/src/main/java/com/team05/petmeeting/domain/user/service/MailService.kt
package com.team05.petmeeting.domain.user.service

import jakarta.mail.internet.MimeMessage
import org.springframework.mail.javamail.JavaMailSender
import org.springframework.mail.javamail.MimeMessageHelper
import org.springframework.stereotype.Service
import org.thymeleaf.TemplateEngine
import org.thymeleaf.context.Context

@Service
class MailService(
    private val mailSender: JavaMailSender,
    private val templateEngine: TemplateEngine,
) {

    fun sendMail(email: String, code: String) {
        runCatching {
            mailSender.createMimeMessage()
                .also { message ->
                    message.writeSignupOtpMail(email, code)
                    mailSender.send(message)
                }
        }.getOrElse {
            throw RuntimeException("메일 발송 실패", it)
        }
    }

    private fun MimeMessage.writeSignupOtpMail(email: String, code: String) {
        val html = templateEngine.process(
            "mail/signupotp",
            Context().apply { setVariable("code", code) },
        )

        MimeMessageHelper(this, true, "UTF-8").apply {
            setTo(email)
            setSubject("Pet Meeting 이메일 인증 코드")
            setText(html, true)
            setFrom("no-reply@petmeeting.com", "PetMeeting")
        }
    }
}
